### PR TITLE
Procatcha support

### PR DESCRIPTION
**Pull request by light-source**
Hello Danny!

I'm Maxim Akimov, a WordPress plugin developer. I collaborate with the [Prosopo](https://prosopo.io/) team and I am responsible for the integration with your plugin. (here is the [related WP support thread](https://wordpress.org/support/topic/proposal-to-add-prosopo-as-a-captcha-vendor-option-10/))

After getting familiar with the plugin architecture, I have done the following:

1. Introduced a new item in your existing integrations for handling the Procaptcha settings
2. Added a new form field type (button) in the JS code (that injects a hidden input into the form)
3. Added the captcha element & API call as a separate class of the newly introduced integration
4. Added the Procaptcha setup status to the general settings (with a link to the integration page)

Let me briefly explain the reasons behind the key decisions:

**1. New integration**

I opted for the new integration as it allows us to keep the vast of the integration code independent, without changing the core code. 

Making all the Procaptcha settings part of the general settings would stretch them and could distract users while adding as part of form settings would require filling the same data multiple times.   

The added 'Procaptcha status' to the general settings simplifies the navigation while keeping the settings short.

**2. Procaptcha field type**

I opted for the button, as unlike auto-injection it allows users not just to enable or disable the captcha for specific forms, but also precisely control its position in the form.

Meanwhile, the field button uses the existing setup and just adds a hidden input. This input is replaced with the real captcha element during rendering. 

It allowed me to avoid big changes in the JS code while having a captcha element dynamic. It means if in the future we've something to change or improve in the element, we won't have the compatibility issue, as it would be in the case with the static HTML.

For the new and modified code, I setup IDE with CodeSniffer to follow your _phpcs.xml_ rules.

Please review the PR, and let me know if you've any notes or concerns. I'm also open to changing the chosen integration approach, in case you don't like it.

Best regards,
Maxim

### CODE REVIEW

**Comment by dannyvankooten** [comment]
Hey @light-source,

Sorry for the delay here - just wanted to let you know that I've cleared up some time today for looking at this in more detail.

File: integrations/prosopo-procaptcha/class-procaptcha.php [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+					)
+				),
+			)
+		);
+
+		if (true === is_wp_error($response)) {
+			// something went wrong, maybe connection issue, but we still shouldn't allow the request.
+			return false;
+		}
+
```
Comment (dannyvankooten) [inline integrations/prosopo-procaptcha/class-procaptcha.php:216]: Should we additionally check the response code here, ie `wp_remote_retrieve_response_code() >= 400`?

File: integrations/prosopo-procaptcha/class-procaptcha.php [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+			// make sure we don't make it twice if other Procaptcha integrations are present.
+			false !== strpos('type="module"', $tag)
+		) {
+			return $tag;
+		}
+
+		// for old WP versions.
+		$tag = str_replace(' type="text/javascript"', '', $tag);
+
+		return str_replace('src', 'type="module" src', $tag);
```
Comment (dannyvankooten) [inline integrations/prosopo-procaptcha/class-procaptcha.php:238]: Just to be safe, I think this should probably have a space prefix:

```php
return str_replace(' src', ' type="module" src', $tag);
```

File: integrations/prosopo-procaptcha/class-procaptcha.php [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+			false === $this->is_displayed_for_authorized &&
+			true === is_user_logged_in() &&
+			false === $is_forced_render
+		) {
+			return '';
+		}
+
+		$this->is_in_use = true;
+
+		$html  = '<mc4wp-procaptcha class="mc4wp-procaptcha" style="display: block;">';
```
Comment (dannyvankooten) [inline integrations/prosopo-procaptcha/class-procaptcha.php:267]: Is this using web components? Neat!

File: includes/views/general-settings.php [hunk 4]
```diff
@@ -65,6 +65,40 @@
...
+							?>
+							<p class="description" style="padding:15px;">
+								<?php
+								printf(
+									// translators: %1$s is the opening <a> tag, %2$s is the closing </a> tag, %3$s is the opening <a> tag, %4$s is the closing </a> tag.
+                                    esc_html__(
+                                        'Click %1$s here %2$s to configure %3$s Procaptcha%4$s, privacy-friendly and GDPR-compliant anti-bot protection.',
+                                        'mailchimp-for-wp'
+                                    ),
+									'<a href="' . esc_html($procaptcha_settings_url) . '">',
```
Comment (dannyvankooten) [inline includes/views/general-settings.php:93]: This should use `esc_attr`

**Review by dannyvankooten — COMMENTED** [review]
Ok, wow, this is really good right away! 

Originally the integrations code was really about adding a sign-up checkbox to forms by other plugins, but I agree that it makes sense to have the Procaptcha specific settings live there.

I may have to work on expanding the forms editor to show a description or help text that links to that specific settings page, just so that users know what the new field is about and how to get it to work.

I left a few minor review comments but once those are taken care of, I think this is good to go!

**Comment by dannyvankooten** [comment]
@light-source I just noticed the general settings section - didn't realize this was on the main settings page. That is a bit too prominent for my liking, so I have removed it in favor of a description on the new field. Sorry about that...

image

I've also tackled a few of my comments, so now only the (probably not so critical) response code check remains. Let me know your thoughts on that one please.

**Comment by light-source** [comment]
Hi @dannyvankooten,

Thank you for the review and for applying the improvements! I'm glad to hear you like the approach.

**1. The general settings**

I hear you and think that the description is a nice workaround to keep it user-friendly. Since your description supports HTML, to improve UX, I would suggest adding a service link, and turning the bolden text into the internal link:

```
description: '<a href="https://prosopo.io/" target="_blank">Procaptcha</a> is a privacy-friendly and GDPR-compliant anti-bot protection. Go to <a target="_blank" href="admin.php?page=mailchimp-for-wp-integrations&integration=prosopo-procaptcha">MC4WP > Integrations > Procaptcha</a> to configure it.'
```

If you don't mind I could make a new PR.

**2. Response code check**

As I see you already added it as well. It makes sense to check the response code rather than a 'cold return' since you have a logger in place. Thank you for adding it.

**Side note: Plugin code quality**

Wanted to express that I was enjoying dealing with your plugin, it's a pleasure to see the right things, like OOP & type declarations in place along with good architecture. Nowadays, many plugins, despite being renowned and used by thousands, have painful code bases.

**Comment by dannyvankooten** [comment]
Perfect @light-source, thanks a ton. I've modified the description so that it now links to the integration settings page instead, good suggestion! 

I think this is good to go out; so we'll be including it in the next plugin release, due on Monday. :tada: 

By the way, although I don't think this is an issue with the code we're discussing here: wheneve I select `pow` as the captchaType, I get an error saying "Incorrect captcha type". The docs state that `pow` is a valid type though, so there must be something going wrong on the prosopo side of things?

**Comment by light-source** [comment]
@dannyvankooten exciting news!

About the captcha type: the local type setting should be consistent with the account type setting, to be the same or stricter. 

E.g. if you've frictionless in the account settings (default), then you can use 'frictionless' (default) or 'image' (stricter) locally. If you want to use 'pure pow', you need to define it on the account level as well.

**Comment by arnelap** [comment]
@light-source I notice recently some CSS code is loaded when the captcha widget is shown on a MC4WP form. It's including some pretty general class names that are often used by themes, such as .widget and .widget__inner  and it seems to break some themes.  Can this be changed to use more unique CSS prefix? Thanks!

**Pull request merged**